design/user-id-update.md

```diff
@@ -0,0 +1,47 @@
+## Make User IDs persistent across session ID, but unique per AppID
+
+This is related to github issue #129.
+
+I want us to redo user IDs and persistence. Current implementation is inconsistent across testbed+Altspace, is inconsistent across registered/unregistered apps, and was designed this way based off of us not having a permission system.
+
+For Background for current implementation, see Issue #129:
+1. Eric wrote: "In a space where an MRE is instantiated multiple times with the same sessionId, a given user in the world will have a different userId in each instance."
+2. Steven wrote this comment : "@eanders-MS This is by design. GDPR requires us to not allow users to be tracked without their permission. As such, any given Altspace user will be assigned a different ID for each instance of a given app. I agree that a persistent ID per app would be tremendously useful, but we have to come up with a strategy to protect our users' privacy first."
+3. Eric wrote about a design:
+	"Registered apps: Same userId for a given user across multiple instances of an app (because the AppId is consistent).
+	Unregistered apps: Different userIds for a given user across multiple instances of an app (because the AppId is derived from the space component id)."
+	"We talked about adding functionality on the website for users to be able to be forgotten by all apps. How this would work: When generating the userId, include a key taken from the user record. When the user chooses to be forgotten, we would roll this key"
+
+Currently in Altspace we use: UserID= combination of altspaceUserID and either appid (for registered apps) or SCID
+
+Currently in testbed we use: UserID=Random generated every connection
+
+### Proposed design
+1. For all host apps have user runtime hash a persistent userID with an AppID or (if AppID isn't available) URL. 
```

Comment:
>URL [](start = 108, length = 3)

It seems to me that URL is going to be *more* persistent than an app ID, not less. Why bother having an app ID at all at that point? Also, if consistent user IDs are going to be the new norm, we should get a privacy review from Legal.

Reply:
I don't agree here - once an app is registered, app ID will be permanent, but sometimes it may be necessary to relocate a server and change a URL. That's the reason we went with app IDs to begin with.
About privacy reviews, we will of course make sure we meet all requirements and recommendations

Reply:
Should we not check with the privacy requirements before making this change?  Seems like a good due diligence step before changing this to establish a persistent user, right?

Reply:
@sorenhan Updating DNS to point to a new server is not hard.

Reply:
While updating DNS is not hard for many situations, there can easily be cases where that's not sufficient without changing URL - for example moving from Heroku or Openode to Azure, from ngrok to a persistent host, or even when running across multiple separate ngrok sessions. Therefore it *has* to be AppID, to avoid any persisted user-specific data to be lost.